chore: add docs

howydev opened the pull request:
1. add docgen output from foundry
2. add smart contract technical documentation

howydev committed:
chore: add docs

jaypaik commented on docs/1-features-overview.md:
```diff
@@ -0,0 +1,18 @@
+# Features Overview
```
Can we surface the stuff under /docs in the root README?

howydev replied:
hmm, how do we do this?

jaypaik replied:
1. Link to these sections within the root README.
2. Just copy paste these sections inline in the root README.

I actually prefer # 2 because most of the time I only look at a repo's root README and nothing else.

howydev replied:
Ok, will make the change. I think as long as we can link to a specific section that'll be fine

jaypaik commented on docs/docgen/src/README.md:
```diff
@@ -0,0 +1,94 @@
+# Modular Account
```
What's the best way to consolidate this so we don't have two copies of the same README?

howydev replied:
oh yeah I can rename this for sure

jaypaik replied:
Do we need this copy for some reason? (I'm not familiar with foundry doc gen stuff)

howydev replied:
name is auto generated also

jaypaik commented on docs/3-security-considerations.md:
Did you miss filling this out?

jaypaik reviewed (commented):
Do we also want to update foundy.toml with the doc settings and do it that way? https://book.getfoundry.sh/reference/config/doc-generator

jaypaik commented on docs/docgen/src/src/README.md:
File structure seems a bit funny with double src. Is that intentional?

howydev replied:
yeah its weird, but its a clean generation hmm. I'll take a deeper look into this

howydev replied:
tried this in other repos and src/src is the normal output lol

howydev committed:
chore: reorg docs to use main readme

howydev committed:
chore: move docs location

howydev committed:
chore: fix link

howydev committed:
chore: remove docgen

howydev committed:
chore: fix pic errors, improve clarity

jaypaik commented on README.md:
````diff
@@ -75,14 +75,92 @@ A deployment script can be found in the `scripts/` folder
 ```
 
-## Security and audits
+## Features Overview
````
Nit: for titles, let's stay consistent and not capitalize each word.

adamegyed commented on README.md:
```diff
@@ -75,14 +75,92 @@ A deployment script can be found in the `scripts/` folder
+Initializer functions are not guarded by any access control modifier. If accounts are not used in a proxy pattern, during the account’s constructor, as per Openzeppelin’s implementation of `Initializable`, initializer functions are able to be reentered. This design choice can be used by an attacker to install additional validations to take over a user’s account.
+
+#### EIP-7702 Accounts and Initializer Functions
+When using EIP-7702, the delegate destinations should only be `SemiModularAccount7702` implementations, and not any of the other account variants. Otherwise, if the delegate destination is set to an account with an `initializer` function, since there isn’t any access control protection on EIP-7702 accounts, an attacker will be able to take over the account.
```
```suggestion
When using EIP-7702, the delegate destination should only be the `SemiModularAccount7702` implementation, and not any of the other account variants. Otherwise, if the delegate destination is set to an account with an unprotected initializer function, an attacker will be able to take over the account.
```

adamegyed commented on README.md:
```diff
@@ -75,14 +75,92 @@ A deployment script can be found in the `scripts/` folder
+
+#### ERC1271 Contract Signatures Support
+
+Certain applications such as Permit2 or Cowswap use the ERC1271 contract signatures standard to determine if a smart contract has approved a certain action. Modular Account v2 implements to allow smart accounts to use these applications.
```
```suggestion
Certain applications such as Permit2 or Cowswap use the ERC1271 contract signatures standard to determine if a smart contract has approved a certain action. Modular Account implements to allow smart accounts to use these applications.
```

Nit: both v1 and v2 support it, but v2 does it natively while v1 has the plugin do it

howydev replied:
Ah good find, when I was writing these docs it was a MA v2 specific doc. But now that it's in this repo it should probably serve both MA v1 and MA v2..

adamegyed commented on README.md:
```diff
@@ -75,14 +75,92 @@ A deployment script can be found in the `scripts/` folder
+
+When modular accounts are created from the factory, an `ERC1967` proxy contract is deployed. Users can update the implementation their proxy points to to choose which smart account implementations to use. Modular Account v2 adheres to the `ERC7201` namespaced storage standard to prevent storage collisions when updating between different implementations.
+
+### Customizing your Modular Account
```
Nit: consistent usage of install vs add in this section

adamegyed commented on README.md:
```diff
@@ -75,14 +75,92 @@ A deployment script can be found in the `scripts/` folder
+2. Installing validations to apply custom validation logic for one or all executions, or uninstalling to remove them
+3. Adding pre validation hooks that are attached to validations, or removing them
+4. Adding execution hooks that are attached to executions, or removing them
+5. Adding execution hooks that are attached to entities, or removing them
```
hmm, seeing this here feels odd when the above section on validation hooks says they're attached to validations. Maybe "validation entities"? Or switch the other one?

adamegyed commented on README.md:
```diff
@@ -75,14 +75,92 @@ A deployment script can be found in the `scripts/` folder
+### Lifecycle of a Runtime Call
+![](./img/runtime-flow.png)
+
+#### Pre-validation Hooks
+
+Pre validation hooks are run before validations. Pre-validation hooks are necessary to perform gas related checks for User Operations (session key gas limits, or gas metering taking into account paymaster usage). These checks must happen in the validation phase since a validation success would allow the entrypoint to charge gas for the user operation to the account.
```
I don't remember the latest on our discussions for "pre-validation hooks" vs just "validation hooks". I think if we go with "validation hooks" here and the sentence still starts with "validation hooks are run before validations", that could be a good enough explainer. Wdyt?

adamegyed commented on README.md:
```diff
@@ -75,14 +75,92 @@ A deployment script can be found in the `scripts/` folder
+
+#### Validations
+
+Validations are usually signature validation functions (BLS, WebAuthn, etc). While it’s feasible to implement signature validation as a pre-validation hook, it’s more efficient and ergonomic to do these in validations since it allows us to apply permissions per entity using execution hooks. In ERC4337, accounts can return validation data that’s not 0 or 1 to signal the usage of a signature aggreg ...
```
```suggestion
Validations are usually signature validation functions (secp256k1, BLS, WebAuthn, etc). While it’s feasible to implement signature validation as a pre-validation hook, it’s more efficient and ergonomic to do these in validations since it allows us to apply permissions per entity using execution hooks. In ERC-4337, accounts can return validation data that’s not 0 or 1 to signal the usage of a signature aggregator.
```

adamegyed commented on README.md:
```diff
@@ -75,14 +75,92 @@ A deployment script can be found in the `scripts/` folder
+#### Execution Hooks
 
-### Bug bounty
+Execution hooks are useful for applying permissions on executions to limit the set of possible actions that can be taken. Post-execution hooks are useful for checking the final state after an execution. Pre and post-execution hook pairs are useful for measuring differences in state due to an execution. For example, you would use a pre and post execution hook pair to enforce that swap outputs from ...
```
```suggestion
Execution hooks are useful for applying permissions on execution functions to limit the set of possible actions that can be taken. Post-execution hooks are useful for checking the final state after an execution. Pre and post-execution hook pairs are useful for measuring differences in state due to an execution. For example, you could use a pre and post execution hook pair to enforce that swap outputs from a DCA swap performed by a session key fall within a some tolerance price determined by a price oracle.
```

adamegyed commented on README.md:
```diff
@@ -75,14 +75,92 @@ A deployment script can be found in the `scripts/` folder
-### Bug bounty
+Execution hooks are useful for applying permissions on executions to limit the set of possible actions that can be taken. Post-execution hooks are useful for checking the final state after an execution. Pre and post-execution hook pairs are useful for measuring differences in state due to an execution. For example, you would use a pre and post execution hook pair to enforce that swap outputs from ...
+
+Execution hooks can be associated either with an (validation module + entity ID) pair to apply permissions on that specific entity, or with an execution selector on the account to apply global restrictions on the account across all entities. A example of a useful global restriction would be to block NFT transfers for NFTs in cold storage, or to apply resource locks.
```
```suggestion
Execution hooks can be associated either with a (validation module + entity ID) pair to apply permissions on that specific entity, or with an execution selector on the account to apply global restrictions on the account across all entities. A example of a useful global restriction would be to block NFT transfers for NFTs in cold storage, or to apply resource locks.
```

adamegyed commented on README.md:
```diff
@@ -75,14 +75,92 @@ A deployment script can be found in the `scripts/` folder
+
+Execution hooks can be associated either with an (validation module + entity ID) pair to apply permissions on that specific entity, or with an execution selector on the account to apply global restrictions on the account across all entities. A example of a useful global restriction would be to block NFT transfers for NFTs in cold storage, or to apply resource locks.
+
+#### Executions
```
Nit: I think this may be confusing to readers, especially since `Execution` has started appearing as the name of a struct holding {address, value, data}. Could we rename this to "execution function" instead?

adamegyed commented on README.md:
```diff
@@ -75,14 +75,92 @@ A deployment script can be found in the `scripts/` folder
+`SemiModularAccountBytecode` (`SMABytecode`) is the cheapest account to deploy. It can only be used for new account deployment, and **should NOT** be used for account upgrades due to requiring specific proxy bytecode.
+
+#### Deferred Actions
+1. In order for a deferred action to be run at validation, in addition to special encoding (which includes the validation to validate the deferred action itself), it must not break ERC-4337 validation-time rules. For instance, this means that any execution hooks on `installValidation` must comply with RIP-7562.
```
```suggestion
1. In order for a deferred action to be run at validation, in addition to special encoding (which includes the validation to validate the deferred action itself), it must not break ERC-4337 validation-time rules. For instance, this means that any execution hooks on `installValidation` must comply with ERC-7562.
```

howydev replied:
TIL - when did this stop being an RIP haha

adamegyed commented on README.md:
```diff
@@ -75,14 +75,92 @@ A deployment script can be found in the `scripts/` folder
+
+#### Deferred Actions
+1. In order for a deferred action to be run at validation, in addition to special encoding (which includes the validation to validate the deferred action itself), it must not break ERC-4337 validation-time rules. For instance, this means that any execution hooks on `installValidation` must comply with RIP-7562.
+2. Deferred actions should only be used to perform actions necessary for user op validation to pass. Otherwise, as deferred actions are not signed over, a malicious bundler could remove the deferred action from the user op and cause an unexpected outcome.
```
This is no longer the case, as the deferred actions flag is now in `userOp.nonce`, so it is signed over

adamegyed reviewed (commented):
Looks great! Did a reasonably thorough read through and had some comments.

howydev committed:
Update README.md

Co-authored-by: Adam Egyed <5456061+adamegyed@users.noreply.github.com>

howydev committed:
Update README.md

Co-authored-by: Adam Egyed <5456061+adamegyed@users.noreply.github.com>

howydev committed:
Update README.md

Co-authored-by: Adam Egyed <5456061+adamegyed@users.noreply.github.com>

howydev committed:
Update README.md

Co-authored-by: Adam Egyed <5456061+adamegyed@users.noreply.github.com>

howydev committed:
chore: review fixes

howydev committed:
chore: more review fixes

howydev committed:
chore: remove old deferred nonce security issue

howydev commented:
@jaypaik @adamegyed made fixes, and also: 
1. Standardized capitalization format across all titles
2. Removed `highlighting` ERC/EIP/RIP variables
3. Standardized usages of "ERC-XXXX" and "ERCXXXX" to always use the first format
4. Switched to using "module entity" language

jaypaik commented on README.md:
```diff
@@ -75,14 +75,91 @@ A deployment script can be found in the `scripts/` folder
 
-Our audit reports can be found in [audits](/audits). The filenames for the reports have the format: `YYYY-MM-DD_VENDOR_FFFFFFF.pdf`, where `YYYY-MM-DD` refers to the date on which the final report was received, `VENDOR` refers to the conductor of the audit, and `FFFFFFF` refers to the short commit hash on which the audit was conducted.
+### Features
+Modular Account v2 can:
```
Nit: I'd change all instances of these
```suggestion
Modular Account can:
```

jaypaik commented on README.md:
```diff
@@ -75,14 +75,91 @@ A deployment script can be found in the `scripts/` folder
+
+### Customizing your Modular Account
+
+The Modular Account v2 can be customized by:
```
```suggestion
Modular Account can be customized by:
```

jaypaik commented on README.md:
```diff
@@ -75,14 +75,91 @@ A deployment script can be found in the `scripts/` folder
+### Lifecycle of a runtime call
+![](./img/runtime-flow.png)
+
+#### Pre-validation hooks
```
Heading hierarchy seems off. Should these be H3?

howydev replied:
These should be under the `Customizing your Modular Account` umbrella since thats where the terms were first introduced, so its the above headers that should be `h4`

jaypaik commented on README.md:
```diff
@@ -75,14 +75,91 @@ A deployment script can be found in the `scripts/` folder
+
+Execution hooks are applied across execution functions. Modular account comes with native execution functions such as `installValidation`, `installExecution`, or `upgradeToAndCall`. However, you could customize the account by installing additional execution functions. After a new execution is installed, when the account is called with that function selector, the account would forward the call to  ...
+
+## Security, Audits, and Bug Bounty
```
```suggestion
## Security
```

jaypaik commented on README.md:
```diff
@@ -75,14 +75,91 @@ A deployment script can be found in the `scripts/` folder
+
+## Security, Audits, and Bug Bounty
+
+Our audit reports can be found in [audits](/audits). The filenames for the reports have the format: `YYYY-MM-DD_VENDOR_FFFFFFF.pdf`, where `YYYY-MM-DD` refers to the date on which the final report was received, `VENDOR` refers to the conductor of the audit, and `FFFFFFF` refers to the short commit hash on which the audit was conducted.
```
```suggestion
### Audits

Our audit reports can be found in [audits](/audits). The filenames for the reports have the format: `YYYY-MM-DD_VENDOR_FFFFFFF.pdf`, where `YYYY-MM-DD` refers to the date on which the final report was received, `VENDOR` refers to the conductor of the audit, and `FFFFFFF` refers to the short commit hash on which the audit was conducted.
```

jaypaik commented on README.md:
```diff
@@ -75,14 +75,91 @@ A deployment script can be found in the `scripts/` folder
+
+Our audit reports can be found in [audits](/audits). The filenames for the reports have the format: `YYYY-MM-DD_VENDOR_FFFFFFF.pdf`, where `YYYY-MM-DD` refers to the date on which the final report was received, `VENDOR` refers to the conductor of the audit, and `FFFFFFF` refers to the short commit hash on which the audit was conducted.
 
 Details of our bug bounty program can be found at https://hackerone.com/alchemyplatform.
```
```suggestion
### Bug bounty program

Details of our bug bounty program can be found at https://hackerone.com/alchemyplatform.
```

jaypaik commented on README.md:
```diff
@@ -75,14 +75,91 @@ A deployment script can be found in the `scripts/` folder
+#### Signature validation flag enablement
+The `isSignatureValidation` flag meant to allow a validation function to validate ERC-1271 signatures. Developer should note that for Modular Account this is a very powerful capability to grant as it allows validation functions to approve deferred actions on the account.
+
 ### Acknowledgements
```
```suggestion
## Acknowledgements
```

howydev committed:
Update README.md

Co-authored-by: Jay Paik <jaypaik@users.noreply.github.com>

howydev committed:
Update README.md

Co-authored-by: Jay Paik <jaypaik@users.noreply.github.com>

howydev committed:
Update README.md

Co-authored-by: Jay Paik <jaypaik@users.noreply.github.com>

howydev committed:
Update README.md

Co-authored-by: Jay Paik <jaypaik@users.noreply.github.com>

howydev committed:
Update README.md

Co-authored-by: Jay Paik <jaypaik@users.noreply.github.com>

howydev committed:
Update README.md

Co-authored-by: Jay Paik <jaypaik@users.noreply.github.com>

howydev committed:
chore: fix header hierarchy
